### PR TITLE
Support custom path for installed extensions

### PR DESCRIPTION
**Pull request by motorina0**
### Summary
Long discussion here: https://github.com/lnbits/lnbits/issues/1579

By default extensions are installed in a hard-coded directory: `lnbits/extensions`.
This PR allow users to specify a custom directory where the extensions will be installed by specifying an [environment variable](https://github.com/lnbits/lnbits/blob/2023c32b9dce0481b865d524b5ebebc1f62433cb/.env.example#L45).

The reason for having a custom extension path is that some users do not want to perform write operations in their `lnbits` directory (maybe they are using Nix which installs programs to immutable locations).

**Required changes for extensions**:
 - this PR is backwards compatible. If the custom extensions path is not used then there is no need for extensions to be updated
 - if the custom extensions path is used then extensions must be updated (see [comment](https://github.com/lnbits/lnbits/pull/1940#issuecomment-1717352903) below)

**Config Steps**:
 - update the `.env` file by configuring `LNBITS_EXTENSIONS_PATH`. This location must allow WRITE operations.
```properties
LNBITS_EXTENSIONS_PATH="/Users/moto/Documents/GitHub/motorina0/lnbits/data/code"
```
 - installed extensions MUST be updated in order to support this custom path feature

### Testing
**Backwards compatible**
 - test that the PR is backwards compatible:
   - do not use the `LNBITS_EXTENSIONS_PATH`
   - install/uninstal/upgrade extensions
 
**Extension Install**
 - set `LNBITS_EXTENSIONS_PATH` to a valid path
 - make sure the extension to be installed has been updated (see [comment](https://github.com/lnbits/lnbits/pull/1940#issuecomment-1717352903) below)
 - install extension
 - check that the extension works as expected
 - upgrade extension and check that the new features are present
 - check that the `lnbits/extensions` directory is empty/does not exist
 - check that the `LNBITS_EXTENSIONS_PATH` has an `extensions` folder that contains the installed extension
 - install older version of the extension and check that the changes are rolled back

**Test data**:
 - manifest file: `https://raw.githubusercontent.com/motorina0/watchonly/main/manifest.json`
 - there are two releases:
    - [v0.1-custom-path-ready](https://github.com/motorina0/watchonly/releases/tag/v0.1-custom-path-ready) - contains the updated paths for the extension
    - [v0.2-ui-api-updates](https://github.com/motorina0/watchonly/releases/tag/v0.2-ui-api-updates) - contains code changes for the extension (see screenshot 1)

### Screenshots
 - 1: UI and API updates
   -  image

### CODE REVIEW

**Comment by motorina0** [comment]
**From**:
```py
tpos_static_files = [
    {
        "path": "/tpos/static",
        "app": StaticFiles(directory="lnbits/extensions/tpos/static"),
        "name": "tpos_static",
    }
]
```
**To**:
```py
tpos_static_files = [
    {
        "path": "/tpos/static",
        "app": "lnbits/extensions/tpos/static",
        "name": "tpos_static",
    }
]
```

**Comment by motorina0** [comment]
Note:
``` log
2023-09-14 09:22:27.51 | WARNING | lnbits.extension_manager:get_installable_extensions:564 | Manifest https://raw.githubusercontent.com/motorina0/watchonly/main/manifest.json failed with 'Invalid version: 'v0.1.3-extern''
```

File: lnbits/app.py [hunk 1]
```diff
@@ -228,9 +230,7 @@ def check_installed_extension_files(ext: InstallableExtension) -> bool:
     if ext.has_installed_version:
         return True
 
-    zip_files = glob.glob(
-        os.path.join(settings.lnbits_data_folder, "extensions", "*.zip")
-    )
+    zip_files = glob.glob(os.path.join(settings.lnbits_data_folder, "zips", "*.zip"))
```
Comment (prusnak) [inline lnbits/app.py:233]: shoudln't this be `lnbits_extensions_path`?

```suggestion
    zip_files = glob.glob(os.path.join(settings.lnbits_extensions_path, "zips", "*.zip"))
```

Comment (motorina0) [reply lnbits/app.py:233]: The idea was that:
  - `settings.lnbits_extensions_path` - is for code
  - `settings.lnbits_data_folder` - is for random files, SQLite files, and (now) zips

**Comment by prusnak** [comment]
It would be great if we could come up with a way how to test whether extensions are compatible with this change in CI of https://github.com/lnbits/lnbits-extensions

Judging by PRs like https://github.com/lnbits/usermanager/pull/14 and https://github.com/lnbits/copilot/pull/9 maybe it is enough to just grep for `StaticFiles(directory="lnbits/extensions/` or `StaticFiles(packages=[("lnbits", "extensions/` and  throw an error if we find a pattern like this? Or should we just disallow using `StaticFiles` in extensions completely?

What do you think @motorina0 @talvasconcelos?

**Comment by prusnak** [comment]
Proposal for a check that detects `StaticFiles` usage in extensions: https://github.com/lnbits/lnbits-extensions/issues/147

**Comment by motorina0** [comment]
Comparability table for "custom extension path support":

| LNbits/Extension | Ext No | Ext Yes |
|------------------|--------|---------|
| **LNbits No**        | OK     | NOK     |
| **LNbits Yes**       | OK     | OK      |

File: lnbits/app.py [hunk 2]
```diff
@@ -270,6 +270,25 @@ def register_routes(app: FastAPI) -> None:
             )
 
 
+def register_custom_extensions_path():
+    if settings.has_default_extension_path:
+        return
+    default_ext_path = "lnbits/extensions"
```
Comment (callebtc) [inline lnbits/app.py:276]: should use `os.path.join()` instead of a string concatenation.

File: lnbits/app.py [hunk 3]
```diff
@@ -270,6 +270,25 @@ def register_routes(app: FastAPI) -> None:
...
+    if os.path.isdir(default_ext_path) and len(os.listdir(default_ext_path)) != 0:
+        logger.warning(
+            "You are using a custom extensions path, "
+            + "but the default extensions directory is not empty. "
+            + f"Please clean-up the '{default_ext_path}' directory."
+        )
+        logger.warning(
+            f"You can move the existing '{default_ext_path}' directory to: "
+            + f" '{settings.lnbits_extensions_path}/extensions'"
```
Comment (callebtc) [inline lnbits/app.py:285]: should use `os.path.join()` instead of a string concatenation.

File: lnbits/app.py [hunk 4]
```diff
@@ -270,6 +270,25 @@ def register_routes(app: FastAPI) -> None:
...
+        )
+
+    sys.path.append(str(Path(settings.lnbits_extensions_path, "extensions")))
+    sys.path.append(str(Path(settings.lnbits_extensions_path, "upgrades")))
```
Comment (callebtc) [inline lnbits/app.py:289]: again os.path.join would help

File: lnbits/app.py (hunk 3 above)
Comment (motorina0) [reply lnbits/app.py:285]: its just a log message

File: lnbits/extension_manager.py [hunk 5]
```diff
@@ -389,21 +400,26 @@ def download_archive(self):
...
     def extract_archive(self):
         logger.info(f"Extracting extension {self.name} ({self.installed_version}).")
-        Path("lnbits", "upgrades").mkdir(parents=True, exist_ok=True)
-        shutil.rmtree(self.ext_upgrade_dir, True)
+        Path(settings.lnbits_extensions_path, "upgrades").mkdir(
+            parents=True, exist_ok=True
+        )
+
+        tmp_dir = Path(settings.lnbits_data_folder, "unzip-temp", self.hash)
+        shutil.rmtree(tmp_dir, True)
```
Comment (callebtc) [inline lnbits/extension_manager.py:408]: I think this one can error (for example insufficient permissions), consider `try: except:`

File: lnbits/extension_manager.py [hunk 6]
```diff
@@ -389,21 +400,26 @@ def download_archive(self):
...
-            Path(self.ext_upgrade_dir, generated_dir_name),
-            Path(self.ext_upgrade_dir, self.id),
+            zip_ref.extractall(tmp_dir)
+        generated_dir_name = os.listdir(tmp_dir)[0]
+        shutil.rmtree(self.ext_upgrade_dir, True)
+        shutil.copytree(
+            Path(tmp_dir, generated_dir_name),
+            Path(self.ext_upgrade_dir),
         )
+        shutil.rmtree(tmp_dir, True)
```
Comment (callebtc) [inline lnbits/extension_manager.py:418]: Same here, all this can error.

File: lnbits/extension_manager.py [hunk 7]
```diff
@@ -389,21 +400,26 @@ def download_archive(self):
...
 
         # Pre-packed extensions can be upgraded
         # Mark the extension as installed so we know it is not the pre-packed version
-        with open(
-            Path(self.ext_upgrade_dir, self.id, "config.json"), "r+"
-        ) as json_file:
+        with open(Path(self.ext_upgrade_dir, "config.json"), "r+") as json_file:
```
Comment (callebtc) [inline lnbits/extension_manager.py:422]: Should you check for existence of this file before opening?

File: lnbits/settings.py [hunk 8]
```diff
@@ -271,10 +271,15 @@ class EnvSettings(LNbitsSettings):
...
     lnbits_title: str = Field(default="LNbits API")
     lnbits_path: str = Field(default=".")
+    lnbits_extensions_path: str = Field(default="./lnbits")
     lnbits_commit: str = Field(default="unknown")
     super_user: str = Field(default="")
     version: str = Field(default="0.0.0")
 
+    @property
+    def has_default_extension_path(self) -> bool:
+        return self.lnbits_extensions_path == "./lnbits"
```
Comment (callebtc) [inline lnbits/settings.py:281]: needs `os.path.join`

**Review by talvasconcelos — APPROVED** [review]

**Review by callebtc — CHANGES_REQUESTED** [review]

File: lnbits/app.py (hunk 4 above)
Comment (motorina0) [reply lnbits/app.py:289]: what is wrong with `Path`?

File: lnbits/extension_manager.py (hunk 5 above)
Comment (motorina0) [reply lnbits/extension_manager.py:408]: in that case it is OK for the exception to be propagated (should not be hidden)

File: lnbits/extension_manager.py (hunk 6 above)
Comment (motorina0) [reply lnbits/extension_manager.py:418]: let it fail, let fail, let it fail ❄️ ☃️ 🏔️

File: lnbits/extension_manager.py (hunk 7 above)
Comment (motorina0) [reply lnbits/extension_manager.py:422]: - if the file does not exist then `open` will throw
- this is fine `extract_archive` should fail in this case

File: lnbits/app.py (hunk 4 above)
Comment (dni) [reply lnbits/app.py:289]: i prefer Path

**Review by dni — APPROVED** [review]
LGTM! awesome

File: lnbits/helpers.py [hunk 9]
```diff
@@ -30,6 +30,10 @@ def url_for(endpoint: str, external: Optional[bool] = False, **params: Any) -> s
 def template_renderer(additional_folders: Optional[List] = None) -> Jinja2Templates:
     folders = ["lnbits/templates", "lnbits/core/templates"]
     if additional_folders:
+        additional_folders += [
+            Path(settings.lnbits_extensions_path, "extensions", f)
```
Comment (prusnak) [inline lnbits/helpers.py:34]: Maybe we want to perform this hack?
```suggestion
            Path(settings.lnbits_extensions_path, "extensions", f[18:] if f.startswith("lnbits/extensions/" else f)
```

**Review by callebtc — APPROVED** [review]
LGTM

**Pull request merged**